src/backend/distributed/commands/table.c

```diff
@@ -700,6 +701,106 @@ PostprocessAlterTableSchemaStmt(Node *node, const char *queryString)
 }
 
 
+/*
+ * SwitchToSequentialAndLocalExecutionIfPrimaryKeyNameTooLong generates the longest primary key name
+ * among the shards of the partitions, and if exceeds the limit switches to sequential and
+ * local execution to prevent self-deadlocks.
+ */
+static void
+SwitchToSequentialAndLocalExecutionIfPrimaryKeyNameTooLong(Oid relationId)
+{
+	if (!PartitionedTable(relationId))
+	{
+		/* Citus already handles long names for regular tables */
+		return;
+	}
+
+	if (ShardIntervalCount(relationId) == 0)
+	{
+		/*
+		 * Relation has no shards, so we cannot run into "long shard index
+		 * name" issue.
+		 */
+		return;
+	}
+
+	Oid longestNamePartitionId = PartitionWithLongestNameRelationId(relationId);
+
+	if (!OidIsValid(longestNamePartitionId))
+	{
+		/* no partitions have been created yet */
+		return;
+	}
+
+	char *longestPartitionShardName = get_rel_name(longestNamePartitionId);
+	ShardInterval *shardInterval = LoadShardIntervalWithLongestShardName(
+		longestNamePartitionId);
+
+	AppendShardIdToName(&longestPartitionShardName, shardInterval->shardId);
+
+
+	Relation rel = RelationIdGetRelation(longestNamePartitionId);
+	Oid namespaceOid = RelationGetNamespace(rel);
+	RelationClose(rel);
+
+	char *primaryKeyName = ChooseIndexName(longestPartitionShardName,
+										   namespaceOid,
+										   NULL, NULL, true, true);
+
+
+	if (primaryKeyName && strnlen(primaryKeyName, NAMEDATALEN) >= NAMEDATALEN - 1)
+	{
+		if (ParallelQueryExecutedInTransaction())
+		{
+			/*
+			 * If there has already been a parallel query executed, the sequential mode
+			 * would still use the already opened parallel connections to the workers,
+			 * thus contradicting our purpose of using sequential mode.
+			 */
+			ereport(ERROR, (errmsg(
```

Comment:
Can we add a test for this? For example, in a transaction block, first run a parallel query then try adding a name-less primary key. This could be useful:
https://github.com/citusdata/citus/blob/main/src/test/regress/sql/multi_index_statements.sql#L315